### ISSUE TITLE
Any way to get kiwix-tools binaries for macOS (or how to build redistributable binaries for macOS)?

### ISSUE THREAD

**Issue by mkathuri** [kiwix/kiwix-tools#464]
Hi there, while looking for kiwix-serve binaries for macOS, I realised they aren't provided yet on the download link (Windows and Linux only). So my first question is, are there any plans to ship kiwix-tools binaries for macOS? If you guys need any help, I can provide help with testing and feedback.

So far, I've tried the docker option and it works, but I would like to avoid it due to the massive battery drain it causes on my Mac. I've also tried using kiwix-build to compile a static version of kiwix-tools (I have no idea how C++ tooling works, and I just need kiwix-serve for my case), but it fails. However the defaults work fine, i.e. it is able to build the native_dyn version, I guess.

The main problem with native_dyn on my system is that the compiled executable isn't redistributable. Hell, if I change the folder structure containing all the compiled artifacts, it fails to resolve the libraries, as it is somehow considering absolute paths for resolving the libraries, which sucks. So here's my second question, am I being stupid and doing something wrong here?

I'm happy to provide more information, if required. I also have Xcode and other developer tools installed, and can do testing as required, I just need help on building redistributable kiwix-tools binaries (preferably static) on macOS.

Thanks for all your hard work. 😄

**Comment by mgautierfr** [comment]
> I've also tried using kiwix-build to compile a static version of kiwix-tools (I have no idea how C++ tooling works, and I just need kiwix-serve for my case), but it fails. 

It would be nice to know how it fails.

We are build every night on MacOs but not static. I don't remember why but there is probably a reason behind.

Another lead could be https://mesonbuild.com/Creating-OSX-packages.html

**Comment by mkathuri** [comment]
Hi @mgautierfr, I followed the instructions of kiwix-build readme (adapted a few bits regarding installing things according to macOS). After finally installing kiwix-build in my own virtualenv, I executed the following command:
`kiwix-build --target-platform native_static`

On the very first line, I get `ERROR: The target platform native_static cannot be build on host Darwin.`

Luckily, the script proceeds anyway, builds and tests a bunch of artefacts, but fails at the last one (kiwix-tools):
```
  build kiwix-tools (native_static):
  configure kiwix-tools : SKIP
  compile kiwix-tools : ERROR
run command 'ninja -v'
current directory is '/Users/maheep/Projects/kiwix-env/kiwix-build/BUILD_native_static/kiwix-tools'
env is :
  TERM_PROGRAM : 'Apple_Terminal'
  SHELL : '/bin/zsh'
  TERM : 'xterm-256color'
  TMPDIR : '/var/folders/kh/hs83qgl54ts_dxym5nb55tth0000gn/T/'
  TERM_PROGRAM_VERSION : '440'
  TERM_SESSION_ID : 'DE116CF2-0451-4ECF-8C99-D8F88611A76C'
  USER : 'maheep'
  SSH_AUTH_SOCK : '/private/tmp/com.apple.launchd.9eksFnSjOq/Listeners'
  PATH : '/Users/maheep/Projects/kiwix-env/kiwix-build/BUILD_native_static/INSTALL/bin:/Users/maheep/Projects/kiwix-env/bin:/opt/local/bin:/opt/local/sbin:/Users/maheep/Library/Python/3.8/bin:/usr/local/bin:/usr/bin:/bin:/usr/sbin:/sbin:/usr/local/aria2/bin:/Library/Apple/usr/bin'
  __CFBundleIdentifier : 'com.apple.Terminal'
  PWD : '/Users/maheep/Projects/kiwix-env/kiwix-build'
  XPC_FLAGS : '0x0'
  XPC_SERVICE_NAME : '0'
  SHLVL : '1'
  HOME : '/Users/maheep'
  LOGNAME : 'maheep'
  OLDPWD : '/Users/maheep/Projects/kiwix-env'
  DISPLAY : ':0'
  VIRTUAL_ENV : '/Users/maheep/Projects/kiwix-env'
  PS1 : '(kiwix-env) %n@%m %1~ %# '
  LC_CTYPE : 'UTF-8'
  _ : '/Users/maheep/Projects/kiwix-env/bin/kiwix-build'
  __CF_USER_TEXT_ENCODING : '0x1F5:0x0:0x0'
  PKG_CONFIG_PATH : ':/Users/maheep/Projects/kiwix-env/kiwix-build/BUILD_native_static/INSTALL/lib/pkgconfig'
  LD_LIBRARY_PATH : ':/Users/maheep/Projects/kiwix-env/kiwix-build/BUILD_native_static/INSTALL/lib:/Users/maheep/Projects/kiwix-env/kiwix-build/BUILD_native_static/INSTALL/lib'
  QMAKE_CXXFLAGS : '-I/Users/maheep/Projects/kiwix-env/kiwix-build/BUILD_native_static/INSTALL/include '
  CPPFLAGS : '-I/Users/maheep/Projects/kiwix-env/kiwix-build/BUILD_native_static/INSTALL/include '
  QMAKE_LFLAGS : '-L/Users/maheep/Projects/kiwix-env/kiwix-build/BUILD_native_static/INSTALL/lib -L/Users/maheep/Projects/kiwix-env/kiwix-build/BUILD_native_static/INSTALL/lib '
  LDFLAGS : '-L/Users/maheep/Projects/kiwix-env/kiwix-build/BUILD_native_static/INSTALL/lib -L/Users/maheep/Projects/kiwix-env/kiwix-build/BUILD_native_static/INSTALL/lib '
[1/4] c++  -o src/reader/kiwix-read 'src/reader/cd3f9f2@@kiwix-read@exe/kiwix-read.cpp.o' -L/Users/maheep/Projects/kiwix-env/kiwix-build/BUILD_native_static/INSTALL/lib -Wl,-dead_strip_dylibs -Wl,-undefined,error -Wl,-headerpad_max_install_names -static-libstdc++ --static /Users/maheep/Projects/kiwix-env/kiwix-build/BUILD_native_static/INSTALL/lib/libkiwix.a /Users/maheep/Projects/kiwix-env/kiwix-build/BUILD_native_static/INSTALL/lib/libzim.a /Users/maheep/Projects/kiwix-env/kiwix-build/BUILD_native_static/INSTALL/lib/liblzma.a -D_THREAD_SAFE -pthread /Users/maheep/Projects/kiwix-env/kiwix-build/BUILD_native_static/INSTALL/lib/libzstd.a /Users/maheep/Projects/kiwix-env/kiwix-build/BUILD_native_static/INSTALL/lib/libicui18n.a /Users/maheep/Projects/kiwix-env/kiwix-build/BUILD_native_static/INSTALL/lib/libicuuc.a /Users/maheep/Projects/kiwix-env/kiwix-build/BUILD_native_static/INSTALL/lib/libicudata.a -lpthread -lm /Users/maheep/Projects/kiwix-env/kiwix-build/BUILD_native_static/INSTALL/lib/libpugixml.a /Users/maheep/Projects/kiwix-env/kiwix-build/BUILD_native_static/INSTALL/lib/libcurl.a /Users/maheep/Projects/kiwix-env/kiwix-build/BUILD_native_static/INSTALL/lib/libz.a /Users/maheep/Projects/kiwix-env/kiwix-build/BUILD_native_static/INSTALL/lib/libmicrohttpd.a /Users/maheep/Projects/kiwix-env/kiwix-build/BUILD_native_static/INSTALL/lib/libxapian.a -L/Users/maheep/Projects/kiwix-env/kiwix-build/BUILD_native_static/INSTALL/lib64 -L/Users/maheep/Projects/kiwix-env/kiwix-build/BUILD_native_static/INSTALL/lib -lidn2
FAILED: src/reader/kiwix-read 
c++  -o src/reader/kiwix-read 'src/reader/cd3f9f2@@kiwix-read@exe/kiwix-read.cpp.o' -L/Users/maheep/Projects/kiwix-env/kiwix-build/BUILD_native_static/INSTALL/lib -Wl,-dead_strip_dylibs -Wl,-undefined,error -Wl,-headerpad_max_install_names -static-libstdc++ --static /Users/maheep/Projects/kiwix-env/kiwix-build/BUILD_native_static/INSTALL/lib/libkiwix.a /Users/maheep/Projects/kiwix-env/kiwix-build/BUILD_native_static/INSTALL/lib/libzim.a /Users/maheep/Projects/kiwix-env/kiwix-build/BUILD_native_static/INSTALL/lib/liblzma.a -D_THREAD_SAFE -pthread /Users/maheep/Projects/kiwix-env/kiwix-build/BUILD_native_static/INSTALL/lib/libzstd.a /Users/maheep/Projects/kiwix-env/kiwix-build/BUILD_native_static/INSTALL/lib/libicui18n.a /Users/maheep/Projects/kiwix-env/kiwix-build/BUILD_native_static/INSTALL/lib/libicuuc.a /Users/maheep/Projects/kiwix-env/kiwix-build/BUILD_native_static/INSTALL/lib/libicudata.a -lpthread -lm /Users/maheep/Projects/kiwix-env/kiwix-build/BUILD_native_static/INSTALL/lib/libpugixml.a /Users/maheep/Projects/kiwix-env/kiwix-build/BUILD_native_static/INSTALL/lib/libcurl.a /Users/maheep/Projects/kiwix-env/kiwix-build/BUILD_native_static/INSTALL/lib/libz.a /Users/maheep/Projects/kiwix-env/kiwix-build/BUILD_native_static/INSTALL/lib/libmicrohttpd.a /Users/maheep/Projects/kiwix-env/kiwix-build/BUILD_native_static/INSTALL/lib/libxapian.a -L/Users/maheep/Projects/kiwix-env/kiwix-build/BUILD_native_static/INSTALL/lib64 -L/Users/maheep/Projects/kiwix-env/kiwix-build/BUILD_native_static/INSTALL/lib -lidn2
clang: warning: argument unused during compilation: '-static-libstdc++' [-Wunused-command-line-argument]
ld: unknown option: --static
clang: error: linker command failed with exit code 1 (use -v to see invocation)
[2/4] c++  -o src/searcher/kiwix-search 'src/searcher/eb86e1d@@kiwix-search@exe/kiwix-search.cpp.o' -L/Users/maheep/Projects/kiwix-env/kiwix-build/BUILD_native_static/INSTALL/lib -Wl,-dead_strip_dylibs -Wl,-undefined,error -Wl,-headerpad_max_install_names -static-libstdc++ --static /Users/maheep/Projects/kiwix-env/kiwix-build/BUILD_native_static/INSTALL/lib/libkiwix.a /Users/maheep/Projects/kiwix-env/kiwix-build/BUILD_native_static/INSTALL/lib/libzim.a /Users/maheep/Projects/kiwix-env/kiwix-build/BUILD_native_static/INSTALL/lib/liblzma.a -D_THREAD_SAFE -pthread /Users/maheep/Projects/kiwix-env/kiwix-build/BUILD_native_static/INSTALL/lib/libzstd.a /Users/maheep/Projects/kiwix-env/kiwix-build/BUILD_native_static/INSTALL/lib/libicui18n.a /Users/maheep/Projects/kiwix-env/kiwix-build/BUILD_native_static/INSTALL/lib/libicuuc.a /Users/maheep/Projects/kiwix-env/kiwix-build/BUILD_native_static/INSTALL/lib/libicudata.a -lpthread -lm /Users/maheep/Projects/kiwix-env/kiwix-build/BUILD_native_static/INSTALL/lib/libpugixml.a /Users/maheep/Projects/kiwix-env/kiwix-build/BUILD_native_static/INSTALL/lib/libcurl.a /Users/maheep/Projects/kiwix-env/kiwix-build/BUILD_native_static/INSTALL/lib/libz.a /Users/maheep/Projects/kiwix-env/kiwix-build/BUILD_native_static/INSTALL/lib/libmicrohttpd.a /Users/maheep/Projects/kiwix-env/kiwix-build/BUILD_native_static/INSTALL/lib/libxapian.a -L/Users/maheep/Projects/kiwix-env/kiwix-build/BUILD_native_static/INSTALL/lib64 -L/Users/maheep/Projects/kiwix-env/kiwix-build/BUILD_native_static/INSTALL/lib -lidn2
FAILED: src/searcher/kiwix-search 
c++  -o src/searcher/kiwix-search 'src/searcher/eb86e1d@@kiwix-search@exe/kiwix-search.cpp.o' -L/Users/maheep/Projects/kiwix-env/kiwix-build/BUILD_native_static/INSTALL/lib -Wl,-dead_strip_dylibs -Wl,-undefined,error -Wl,-headerpad_max_install_names -static-libstdc++ --static /Users/maheep/Projects/kiwix-env/kiwix-build/BUILD_native_static/INSTALL/lib/libkiwix.a /Users/maheep/Projects/kiwix-env/kiwix-build/BUILD_native_static/INSTALL/lib/libzim.a /Users/maheep/Projects/kiwix-env/kiwix-build/BUILD_native_static/INSTALL/lib/liblzma.a -D_THREAD_SAFE -pthread /Users/maheep/Projects/kiwix-env/kiwix-build/BUILD_native_static/INSTALL/lib/libzstd.a /Users/maheep/Projects/kiwix-env/kiwix-build/BUILD_native_static/INSTALL/lib/libicui18n.a /Users/maheep/Projects/kiwix-env/kiwix-build/BUILD_native_static/INSTALL/lib/libicuuc.a /Users/maheep/Projects/kiwix-env/kiwix-build/BUILD_native_static/INSTALL/lib/libicudata.a -lpthread -lm /Users/maheep/Projects/kiwix-env/kiwix-build/BUILD_native_static/INSTALL/lib/libpugixml.a /Users/maheep/Projects/kiwix-env/kiwix-build/BUILD_native_static/INSTALL/lib/libcurl.a /Users/maheep/Projects/kiwix-env/kiwix-build/BUILD_native_static/INSTALL/lib/libz.a /Users/maheep/Projects/kiwix-env/kiwix-build/BUILD_native_static/INSTALL/lib/libmicrohttpd.a /Users/maheep/Projects/kiwix-env/kiwix-build/BUILD_native_static/INSTALL/lib/libxapian.a -L/Users/maheep/Projects/kiwix-env/kiwix-build/BUILD_native_static/INSTALL/lib64 -L/Users/maheep/Projects/kiwix-env/kiwix-build/BUILD_native_static/INSTALL/lib -lidn2
clang: warning: argument unused during compilation: '-static-libstdc++' [-Wunused-command-line-argument]
ld: unknown option: --static
clang: error: linker command failed with exit code 1 (use -v to see invocation)
[3/4] c++  -o src/manager/kiwix-manage 'src/manager/2de60dd@@kiwix-manage@exe/kiwix-manage.cpp.o' -L/Users/maheep/Projects/kiwix-env/kiwix-build/BUILD_native_static/INSTALL/lib -Wl,-dead_strip_dylibs -Wl,-undefined,error -Wl,-headerpad_max_install_names -static-libstdc++ --static /Users/maheep/Projects/kiwix-env/kiwix-build/BUILD_native_static/INSTALL/lib/libkiwix.a /Users/maheep/Projects/kiwix-env/kiwix-build/BUILD_native_static/INSTALL/lib/libzim.a /Users/maheep/Projects/kiwix-env/kiwix-build/BUILD_native_static/INSTALL/lib/liblzma.a -D_THREAD_SAFE -pthread /Users/maheep/Projects/kiwix-env/kiwix-build/BUILD_native_static/INSTALL/lib/libzstd.a /Users/maheep/Projects/kiwix-env/kiwix-build/BUILD_native_static/INSTALL/lib/libicui18n.a /Users/maheep/Projects/kiwix-env/kiwix-build/BUILD_native_static/INSTALL/lib/libicuuc.a /Users/maheep/Projects/kiwix-env/kiwix-build/BUILD_native_static/INSTALL/lib/libicudata.a -lpthread -lm /Users/maheep/Projects/kiwix-env/kiwix-build/BUILD_native_static/INSTALL/lib/libpugixml.a /Users/maheep/Projects/kiwix-env/kiwix-build/BUILD_native_static/INSTALL/lib/libcurl.a /Users/maheep/Projects/kiwix-env/kiwix-build/BUILD_native_static/INSTALL/lib/libz.a /Users/maheep/Projects/kiwix-env/kiwix-build/BUILD_native_static/INSTALL/lib/libmicrohttpd.a /Users/maheep/Projects/kiwix-env/kiwix-build/BUILD_native_static/INSTALL/lib/libxapian.a -L/Users/maheep/Projects/kiwix-env/kiwix-build/BUILD_native_static/INSTALL/lib64 -L/Users/maheep/Projects/kiwix-env/kiwix-build/BUILD_native_static/INSTALL/lib -lidn2
FAILED: src/manager/kiwix-manage 
c++  -o src/manager/kiwix-manage 'src/manager/2de60dd@@kiwix-manage@exe/kiwix-manage.cpp.o' -L/Users/maheep/Projects/kiwix-env/kiwix-build/BUILD_native_static/INSTALL/lib -Wl,-dead_strip_dylibs -Wl,-undefined,error -Wl,-headerpad_max_install_names -static-libstdc++ --static /Users/maheep/Projects/kiwix-env/kiwix-build/BUILD_native_static/INSTALL/lib/libkiwix.a /Users/maheep/Projects/kiwix-env/kiwix-build/BUILD_native_static/INSTALL/lib/libzim.a /Users/maheep/Projects/kiwix-env/kiwix-build/BUILD_native_static/INSTALL/lib/liblzma.a -D_THREAD_SAFE -pthread /Users/maheep/Projects/kiwix-env/kiwix-build/BUILD_native_static/INSTALL/lib/libzstd.a /Users/maheep/Projects/kiwix-env/kiwix-build/BUILD_native_static/INSTALL/lib/libicui18n.a /Users/maheep/Projects/kiwix-env/kiwix-build/BUILD_native_static/INSTALL/lib/libicuuc.a /Users/maheep/Projects/kiwix-env/kiwix-build/BUILD_native_static/INSTALL/lib/libicudata.a -lpthread -lm /Users/maheep/Projects/kiwix-env/kiwix-build/BUILD_native_static/INSTALL/lib/libpugixml.a /Users/maheep/Projects/kiwix-env/kiwix-build/BUILD_native_static/INSTALL/lib/libcurl.a /Users/maheep/Projects/kiwix-env/kiwix-build/BUILD_native_static/INSTALL/lib/libz.a /Users/maheep/Projects/kiwix-env/kiwix-build/BUILD_native_static/INSTALL/lib/libmicrohttpd.a /Users/maheep/Projects/kiwix-env/kiwix-build/BUILD_native_static/INSTALL/lib/libxapian.a -L/Users/maheep/Projects/kiwix-env/kiwix-build/BUILD_native_static/INSTALL/lib64 -L/Users/maheep/Projects/kiwix-env/kiwix-build/BUILD_native_static/INSTALL/lib -lidn2
clang: warning: argument unused during compilation: '-static-libstdc++' [-Wunused-command-line-argument]
ld: unknown option: --static
clang: error: linker command failed with exit code 1 (use -v to see invocation)
[4/4] c++  -o src/server/kiwix-serve 'src/server/e19e62e@@kiwix-serve@exe/kiwix-serve.cpp.o' -L/Users/maheep/Projects/kiwix-env/kiwix-build/BUILD_native_static/INSTALL/lib -Wl,-dead_strip_dylibs -Wl,-undefined,error -Wl,-headerpad_max_install_names -static-libstdc++ --static /Users/maheep/Projects/kiwix-env/kiwix-build/BUILD_native_static/INSTALL/lib/libkiwix.a /Users/maheep/Projects/kiwix-env/kiwix-build/BUILD_native_static/INSTALL/lib/libzim.a /Users/maheep/Projects/kiwix-env/kiwix-build/BUILD_native_static/INSTALL/lib/liblzma.a -D_THREAD_SAFE -pthread /Users/maheep/Projects/kiwix-env/kiwix-build/BUILD_native_static/INSTALL/lib/libzstd.a /Users/maheep/Projects/kiwix-env/kiwix-build/BUILD_native_static/INSTALL/lib/libicui18n.a /Users/maheep/Projects/kiwix-env/kiwix-build/BUILD_native_static/INSTALL/lib/libicuuc.a /Users/maheep/Projects/kiwix-env/kiwix-build/BUILD_native_static/INSTALL/lib/libicudata.a -lpthread -lm /Users/maheep/Projects/kiwix-env/kiwix-build/BUILD_native_static/INSTALL/lib/libpugixml.a /Users/maheep/Projects/kiwix-env/kiwix-build/BUILD_native_static/INSTALL/lib/libcurl.a /Users/maheep/Projects/kiwix-env/kiwix-build/BUILD_native_static/INSTALL/lib/libz.a /Users/maheep/Projects/kiwix-env/kiwix-build/BUILD_native_static/INSTALL/lib/libmicrohttpd.a /Users/maheep/Projects/kiwix-env/kiwix-build/BUILD_native_static/INSTALL/lib/libxapian.a -L/Users/maheep/Projects/kiwix-env/kiwix-build/BUILD_native_static/INSTALL/lib64 -L/Users/maheep/Projects/kiwix-env/kiwix-build/BUILD_native_static/INSTALL/lib -lidn2
FAILED: src/server/kiwix-serve 
c++  -o src/server/kiwix-serve 'src/server/e19e62e@@kiwix-serve@exe/kiwix-serve.cpp.o' -L/Users/maheep/Projects/kiwix-env/kiwix-build/BUILD_native_static/INSTALL/lib -Wl,-dead_strip_dylibs -Wl,-undefined,error -Wl,-headerpad_max_install_names -static-libstdc++ --static /Users/maheep/Projects/kiwix-env/kiwix-build/BUILD_native_static/INSTALL/lib/libkiwix.a /Users/maheep/Projects/kiwix-env/kiwix-build/BUILD_native_static/INSTALL/lib/libzim.a /Users/maheep/Projects/kiwix-env/kiwix-build/BUILD_native_static/INSTALL/lib/liblzma.a -D_THREAD_SAFE -pthread /Users/maheep/Projects/kiwix-env/kiwix-build/BUILD_native_static/INSTALL/lib/libzstd.a /Users/maheep/Projects/kiwix-env/kiwix-build/BUILD_native_static/INSTALL/lib/libicui18n.a /Users/maheep/Projects/kiwix-env/kiwix-build/BUILD_native_static/INSTALL/lib/libicuuc.a /Users/maheep/Projects/kiwix-env/kiwix-build/BUILD_native_static/INSTALL/lib/libicudata.a -lpthread -lm /Users/maheep/Projects/kiwix-env/kiwix-build/BUILD_native_static/INSTALL/lib/libpugixml.a /Users/maheep/Projects/kiwix-env/kiwix-build/BUILD_native_static/INSTALL/lib/libcurl.a /Users/maheep/Projects/kiwix-env/kiwix-build/BUILD_native_static/INSTALL/lib/libz.a /Users/maheep/Projects/kiwix-env/kiwix-build/BUILD_native_static/INSTALL/lib/libmicrohttpd.a /Users/maheep/Projects/kiwix-env/kiwix-build/BUILD_native_static/INSTALL/lib/libxapian.a -L/Users/maheep/Projects/kiwix-env/kiwix-build/BUILD_native_static/INSTALL/lib64 -L/Users/maheep/Projects/kiwix-env/kiwix-build/BUILD_native_static/INSTALL/lib -lidn2
clang: warning: argument unused during compilation: '-static-libstdc++' [-Wunused-command-line-argument]
ld: unknown option: --static
clang: error: linker command failed with exit code 1 (use -v to see invocation)
ninja: build stopped: subcommand failed.


Stopping build due to errors
```

Hopefully this error message should help you arrive at the solution. Please reach out in case more details are needed.

**Comment by mkathuri** [comment]
Posting a new comment, because it seems I've managed to partially solve my problem after some thorough reading from the following sources:
- [Statically linked binaries on Mac OS X](https://developer.apple.com/library/archive/qa/qa1118/_index.html)
- [Wading through macOS Linker hell](https://gist.github.com/loderunner/b6846dd82967ac048439)

I replaced the last failing command (the one for kiwix-serve) with this:
```
c++ -o src/server/kiwix-serve 'src/server/e19e62e@@kiwix-serve@exe/kiwix-serve.cpp.o' -L/Users/maheep/Projects/kiwix-env/kiwix-build/BUILD_native_static/INSTALL/lib -Wl,-dead_strip_dylibs -Wl,-undefined,error -Wl,-headerpad_max_install_names -D_THREAD_SAFE -pthread -L/Users/maheep/Projects/kiwix-env/kiwix-build/BUILD_native_static/INSTALL/lib64 -L/Users/maheep/Projects/kiwix-env/kiwix-build/BUILD_native_static/INSTALL/lib -lidn2 -lkiwix -lzim -llzma -lzstd -licui18n -licuuc -licudata -lpthread -lm -lpugixml -lcurl -lz -lmicrohttpd -lxapian
```

The short version is this: You shouldn't tell macOS to make a statically linked binary, period. You can have statically linked libraries in your binary, but the binary itself should never be statically linked. That's why --static option won't work. I think. That and it doesn't exist anyway on macOS. **Also, if macOS finds both static and dynamic libs in a folder, it'll always prefer dynamic libs. BE CAREFUL.**

All the prequisites except **libidn2** already existed in the INSTALL/lib folder, I tried building libidn2 as a static library and linking against it, but ran into problems. So, here's what I did:

- Downloaded latest source of libidn2 from: https://ftp.gnu.org/gnu/libidn/libidn2-latest.tar.gz
- `./configure && sudo make install`. It adds the libraries to my /usr/share/lib folder. I tried it with --enable-static as well but it didn't work for me. Or maybe I'm too tired. I hope someone can get this working.
- Run kiwix-build with static options. It fails at the last steps (as seen in above logs).
- Replace the failing command with the one mentioned above.

**Caveats**:
- I'm still dynamically linked to libidn2. And I need it installed in my /usr/local/lib folder.
- No idea about backwards compatibility.

**Pros**:
- I can finally move the executable around and safely remove the build directory. Yippy!

**Comment by kelson42** [comment]
IMO this ticket is a duplicate of #454. The main reason why this is not already implement is lack of time/prio. But we would love to provide them.

**Comment by mkathuri** [comment]
Happy to report that I'm able to build all of the tools statically now. I've printed the general --help output of every tool and everything seems to be in order on my system running macOS Big Sur 11.4. No claims of backward compatibility are made, and the binaries are provided as is. Since kiwix-build is yet to be fixed, I had to manually compile each tool and create the zip file by hand. Here's the link if anyone wants to test: https://www.icloud.com/iclouddrive/0cST4FGV4otbUeASVyzpr69RA#kiwix-tools-bin

Both kiwix-serve and kiwix-manage need to link against iconv as well, using -liconv. The other two tools don't need it.

**Comment by kelson42** [comment]
@rgaudin @mgautierfr would you be interested to test this and bring this in kiwix-build?

**Comment by rgaudin** [comment]
@kelson42 thanks for pinging me ; I wasn't subscribed to this. There never really was any blocker in building kiwix-serve on macOS. It's just that this is not integrated in kiwix-build because it's a bit complicated and different than on Linux.

It's great that @mkathuri got it working but:

- we plan on building kiwix tools around libkiwix, not each tool individually. My personal kiwix-serve build is only linked to libkiwix for instance.
- most of the work is to be done on libkiwix compilation for macOS integration
- Same amount of work is still required to integrate it in kiwix-build so this doesn't advances https://github.com/kiwix/kiwix-build/issues/454 much

**Comment by mkathuri** [comment]
@rgaudin, I'm writing this just for the sake of gaining a better understanding of your points:
> we plan on building kiwix tools around libkiwix, not each tool individually. My personal kiwix-serve build is only linked to libkiwix for instance.

Presently for other OSes (Windows and Linux), you guys ship fat binaries, right? Per my understanding of the above comments, you guys are planning to change this and ship out a libkiwix dll of sorts with all your binaries in the future - which means all executables get linked dynamically to just the libkiwix library, which is statically linked to various other libs (eg. zlib, libcurl) and dynamically, of course to system-shipped libraries (libstd, VC++ runtime etc).

If such is the case, I must say its a great approach, since I think it'll be widely supported across different OSes.

> most of the work is to be done on libkiwix compilation for macOS integration

If I am able to compile kiwix-tools on my Mac, doesn't it mean that libkiwix already builds on my machine? Just trying to understand what work needs to be done, and if I can help.

Last point is clear.

Thanks for your comments. :)

**Comment by mgautierfr** [comment]
Thanks @mkathuri for this investigation.


> The short version is this: You shouldn't tell macOS to make a statically linked binary, period. You can have statically linked libraries in your binary, but the binary itself should never be statically linked. That's why --static option won't work. I think. That and it doesn't exist anyway on macOS. Also, if macOS compiler finds both static and dynamic libs in a folder, it'll always prefer dynamic libs. BE CAREFUL.

I've you tried the target `native_mixed` ?
What it does is to compile all dependencies statically and the "final" project (kiwix-tools by default) dynamically. It should globally do as we need to do.

> These steps, if correct, have to be incorporated in kiwix-build accordingly (I have no idea how to do that).

Yes, in any way these steps must be integrated in project build system (kiwix-tools meson file, ...) or at least in kiwix-build. We must not have those commands to be run by hand by the user. If you want to do the PR, I can help you to incorporate this in kiwix-build.


> Presently for other OSes (Windows and Linux), you guys ship fat binaries, right? Per my understanding of the above comments, you guys are planning to change this and ship out a libkiwix dll of sorts with all your binaries in the future - which means all executables get linked dynamically to just the libkiwix library, which is statically linked to various other libs (eg. zlib, libcurl) and dynamically, of course to system-shipped libraries (libstd, VC++ runtime etc).
>
> If such is the case, I must say its a great approach, since I think it'll be widely supported across different OSes.

No. We don't have such plan. We build our tools statically and distribute it this way. Of course, other distributor (distrubution packaging system, ...) may (and do) build them dynamically but, us, we are distributing "simple" binaries [*]

[*] As far as possible. kiwix-desktop on windows for instance is a archive to decompress with a lot of dll in it. On linux it is either a appimage (a simple binary from user pov or a flatpak)

@rgaudin I'm not sur to totally understand all your points :

> we plan on building kiwix tools around libkiwix, not each tool individually. My personal kiwix-serve build is only linked to libkiwix for instance.

This is not the case. For example we will not integrate docopt in libkiwix. This is a library for command line tools only and it will stay this way.
If somehow tools need a specific dependency it will link with it (whatever if libkiwix is already using it or not)

> most of the work is to be done on libkiwix compilation for macOS integration

It seems that libkiwix is already compiling on macOsS

> Same amount of work is still required to integrate it in kiwix-build so this doesn't advances kiwix/kiwix-build#454 much

It is good to know what have to be done before starting the integration in kiwix-build. And kiwix/kiwix-build#454 takes another approche : build a DMG instead of a "static" binary.

**Comment by rgaudin** [comment]
No sorry I didn't make myself clear. What I meant is that most of the compilation effort is to be done in kiwix-build for `libkiwix`. We don't want to link libzim to kiwix-serve for example, as it'd be in libkiwix. @mgautierfr seems to say the opposite but I guess he's referring to direct in-tool usage vs usage through libkiwix. We're on the same page.

We need to publish libkiwix in a standalone and reusable manner (https://github.com/kiwix/kiwix-build/issues/455) for iOS/macOS desktop and we also need to produce stages that can be bundled into the native_mixed cli tools that needs to be published as well: zim-tools and kiwix-tools.

AFAIK the only macOS deliverable ATM is libzim.

> And kiwix/kiwix-build#454 takes another approche : build a DMG instead of a "static" binary.

DMG is a container format so I'm not sure how relevant this is here…

**Comment by mgautierfr** [comment]
> DMG is a container format so I'm not sure how relevant this is here…

I think we agree here. We are both saying that this is two different issues and doing one issue will not help the other.
(Except they have both the same finality : publish a executable binary)


> @mgautierfr seems to say the opposite but I guess he's referring to direct in-tool usage vs usage through libkiwix. We're on the same page.

I think we agree on the theory, but not sure we agree on what we will do. I will try to be more exhaustive in the end of this comment.

---

TLDR : The following is a bit of subject as this issue is about distributing binary of the executables kiwix-tools.
But I'm doing  it anyway as we need to be clear on this

If `libkiwix` uses `libzim` "internally" and `kiwix-tools` uses `libkiwix` "only", then `libzim` is a dependency of `libkiwix` not of `kiwix-tools`.

The standard way of building/linking is :
- If we are compiling dynamicly, we "build" only with direct dependencies. But we need to distribute all .so and at runtime we link with all dependencies.
- If we are compiling statically , we build with all dependencies.

When we distribute libraries, we need to distribute a .so but in the same time we want to distribute only one binary. So we build all dependencies statically and we build the library as a dynamic library (.so) but with static linking (we integrate at build time the dependencies).

So we have a .so which contains almost everything. But the almost is important. As we link statically with the dependency only the code really used is keep. If libkiwix does not use the function `zim::Foo`, it will not be in the library.
If `kiwix-tools` start to directly use libzim, it may want to use `zim::Foo`. So it is not enough to link with the "fat" `libkiwix.so`. We also need to make kiwix-tools explicitly compile with libzim.so.

But then we may compile `libkiwix.so` with one version of libzim and kiwix-tools with another one. With potential conflict (https://github.com/kiwix/kiwix-build/issues/418#issuecomment-610330985).
This is why distributing prebuild library is rarely a good idea. (At least not half static binaries). Either we build everything dynamic and let resolve linking at runtime. Or we build everything static and we resolve linking at compilation time (of the final executable).

And with https://github.com/kiwix/libkiwix/issues/430, kiwix-tools will NEED to directly use libzim.

**Comment by rgaudin** [comment]
Thanks for the detailed explanation ; we're on the same page 👍

**Comment by mkathuri** [comment]
@mgautierfr, @rgaudin thank you for your comments. It seems the approach I've taken is going well so far. I'll try building and running against older macOS versions to see if compatibility is/can be maintained.

@mgautierfr regarding integrating these steps in project build system or kiwix-build, I'm afraid I lack the requisite knowledge of both meson and python, and since I'm mostly busy with my 24x7 job, I won't be able to do it. I am happy to test out any build procedure or binaries, however - now and in the future 👍.

I believe my issue is now resolved (ended up building binaries myself, 😄). Thank you all for your help and hard work.

**Comment by kelson42** [comment]
@mkathuri thx for the transparency and feedback on this.
